Move WPT to Authorization header (#287)
ietf-wg-wimse/draft-ietf-wimse-s2s-protocol: draft-ietf-wimse-wpt.md

# Workload Proof Token  {#wpt}

The Workload Proof Token (WPT) is a JWT that provides proof of possession of the private key associated with a Workload Identity Token (WIT). The Workload Identity Token is sent in the request as described in {{!I-D.ietf-wimse-workload-creds}}. The WPT is sent in the `Authorization` header field ({{Section 11.6.2 of RFC9110}}) of the request, using the `WPT` authentication scheme defined by this document.
The ABNF {{RFC5234}} syntax of the credentials of the `WPT` authentication scheme is shown in {{wpt-scheme-abnf}} below.
As with any HTTP authentication scheme, the scheme name is case-insensitive ({{Section 11.1 of RFC9110}}).

~~~ abnf
ALPHA = %x41-5A / %x61-7A ; A-Z / a-z
DIGIT = %x30-39 ; 0-9
SP = %x20 ; space
base64url = 1*(ALPHA / DIGIT / "-" / "_")
JWT =  base64url "." base64url "." base64url

> credentials = "WPT" 1*SP JWT

Author: @bc-pi you have much more experience here, can you take a closer look at this abnf update please?

Reviewer: This is probably okay but with the move to the Authorization header I think a lot of this abnf isn't needed anymore (if it ever was "needed"). 

I had something more like the top bits of https://datatracker.ietf.org/doc/html/rfc9449#name-the-dpop-authentication-sch in mind. Where the abnf is really just restating/showing how the new scheme and token fit in the HTTP auth framework.

Reviewer: I think this section could be adapted towards that end without too much difficulty.

Author: Updated.

# Workload Proof Token  {#wpt}

The Workload Proof Token (WPT) is a JWT that provides proof of possession of the private key associated with a Workload Identity Token (WIT). The Workload Identity Token is sent in the request as described in {{!I-D.ietf-wimse-workload-creds}}. The WPT is sent in the `Authorization` header field ({{Section 11.6.2 of RFC9110}}) of the request, using the `WPT` authentication scheme defined by this document.
The syntax of the credentials of the `WPT` authentication scheme uses the `token68` syntax
defined in {{Section 11.2 of RFC9110}}, whose ABNF {{RFC5234}} is repeated in {{wpt-scheme-abnf}} below
for ease of reference.
As with any HTTP authentication scheme, the scheme name is case-insensitive ({{Section 11.1 of RFC9110}}),
while case is significant in the credentials.

~~~ abnf
token68     = 1*( ALPHA / DIGIT /
                  "-" / "." / "_" / "~" / "+" / "/" ) *"="

> credentials = "WPT" 1*SP token68

Reviewer 2: I don't see a reason NOT to use the `auth-param` syntax, i.e.

```
Authorization: WPT token="<jwt>"
```

which RFC 9110 sec. 16.4.2 says we ought to do. It's a bit uglier and a bit more extensible, and we stay on the good side of the HTTP community.

To validate the WPT in the request, the recipient MUST ensure the following:

* There is exactly one `Authorization` header field in the request and it uses the `WPT` authentication scheme.

> * The credentials of the `WPT` authentication scheme are a single and well-formed JWT, as per {{wpt-scheme-abnf}}.

Reviewer: If the changes I suggested / implied in https://github.com/ietf-wg-wimse/draft-ietf-wimse-s2s-protocol/pull/287/changes#r3799048449 happen this ref prob isn't quite right. 
```suggestion
* The credentials of the `WPT` authentication scheme are a single and well-formed JWT.
```

Author: {{wpt-scheme-abnf}} is still there so I think it still applies?

Reviewer: {{wpt-scheme-abnf}} is still there but doesn't add anything useful to this step. A a single and well-formed JWT is valid token68 but being a valid token68 thing doesn't tell you anything more of value and there are tons of valid token68 things that aren't well-formed JWTs.

To validate the WPT in the request, the recipient MUST first validate the WIT as specified in {{Section 5.1.4 of I-D.ietf-wimse-workload-creds}}, then ensure the following:

> * There is exactly one `Authorization` header field in the request and it uses the `WPT` authentication scheme.

Reviewer 2: A separate PR mentions that we need to validate the WIT, too. Please make sure you don't lose it.

Reviewer: the line right above this one?

## Error Conditions

Errors may occur during the processing of the WPT. If the validation fails for any reason,
such as an invalid signature, an expired validity time window, or a malformed data structure, an error is returned.
Because the WPT is conveyed with an HTTP authentication scheme, a recipient that rejects the request for a reason
related to the WPT SHOULD use the HTTP status code 401 (Unauthorized). Such a response carries a `WWW-Authenticate`
header field with at least one challenge, as required by {{Section 11.6.1 of RFC9110}}, and a recipient that accepts
WPTs includes a challenge with the `WPT` scheme. This specification defines no authentication parameters
for the `WPT` challenge. An example of such a response header field is shown below.

~~~ http
WWW-Authenticate: WPT
~~~

> {: title="Example WPT Challenge"}

Reviewer 2: Are you sure the name "challenge" still applies? I have always thought that the name refers to a random nonce included in the www-authenticate header.

Reviewer 2: sorry, my mistake. Closing.

WPTs includes a challenge with the `WPT` scheme. This specification defines no authentication parameters
for the `WPT` challenge. An example of such a response header field is shown below.

~~~ http
WWW-Authenticate: WPT
~~~
{: title="Example WPT Challenge"}

An HTTP status code such as 400 (Bad Request) remains appropriate for errors that are not related to the presented
credentials. Either response could include more details as per {{RFC9457}}, such as an indicator that the wrong key
material or algorithm was used.

## Coexistence with Bearer Tokens {#coexist}

The WPT occupies the `Authorization` header field of the request. That field is defined as
`Authorization = credentials` ({{Section 11.6.2 of RFC9110}}); its value is a single set of credentials rather than a
comma-separated list, and a sender therefore MUST NOT generate more than one `Authorization` header field
({{Section 5.3 of RFC9110}}). A request that presents a WPT consequently cannot present a token with any other
authentication scheme, including the `Bearer` scheme ({{Section 2.1 of RFC6750}}) or the `DPoP` scheme

> ({{Section 7.1 of RFC9449}}).

Reviewer: It feels a little weird to me to talk about DPoP in a section called Coexistence with Bearer Tokens. 
```suggestion
authentication scheme, including the `Bearer` scheme ({{Section 2.1 of RFC6750}}) or the `Basic` scheme ({{RFC7617}}).
```

## Error Conditions

Errors may occur during the processing of the WPT. If the signature verification fails for any reason,
such as an invalid signature, an expired validity time window, or a malformed data structure, an error is returned.
Because the WPT is conveyed with an HTTP authentication scheme, a recipient that rejects the request for a reason
related to the WPT SHOULD use the HTTP status code 401 (Unauthorized). Such a response carries a `WWW-Authenticate`
header field with at least one challenge, as required by {{Section 11.6.1 of RFC9110}}, and a recipient that accepts
WPTs includes a challenge with the `WPT` scheme. This specification defines no authentication parameters

> for the `WPT` challenge. An example of such a response header field is shown below.

Reviewer: did we want to consider parameter(s) to indicate acceptable audience value(s)? Although, even as I type that prior sentence, I start to have doubts about that would really work...

Reviewer: meh

~~~ http
WWW-Authenticate: WPT
~~~
{: title="Example WPT Challenge"}

An HTTP status code such as 400 (Bad Request) remains appropriate for errors that are not related to the presented
credentials. Either response could include more details as per {{RFC9457}}, such as an indicator that the wrong key
material or algorithm was used.

## Coexistence with Bearer Tokens {#coexist}

The WPT occupies the `Authorization` header field of the request. That field is defined as
`Authorization = credentials` ({{Section 11.6.2 of RFC9110}}); its value is a single set of credentials rather than a
comma-separated list, and a sender therefore MUST NOT generate more than one `Authorization` header field
({{Section 5.3 of RFC9110}}). A request that presents a WPT consequently cannot present a token with any other

> authentication scheme, including the `Bearer` scheme ({{Section 2.1 of RFC6750}}) and the `DPoP` scheme

Reviewer: I think... 
```suggestion
authentication scheme, including the `Bearer` scheme ({{Section 2.1 of RFC6750}}) or the `DPoP` scheme
```

authorization decisions about the caller. A bearer token can be used by anyone who obtains it, so honoring one
alongside a WPT would reduce the strength of the request to that of its weakest credential and defeat the proof of
possession that the WIT and WPT provide.

Tokens that convey end-user identity or authorization context rather than the identity of the calling workload are
not affected, provided that they are conveyed in their own header fields. Transaction tokens
{{?I-D.ietf-oauth-transaction-tokens}} are an example: they are carried in the `Txn-Token` header field and are
explicitly not permitted in the `Authorization` header field, precisely so that the field remains available for
authenticating the call itself. Such tokens are bound to the WPT with the `tth` and `oth` claims described in
{{wpt}}. In this case a workload's
authorization policy may take into account both the sending workload's identity and the information in the context token. For example, the
identity in the WIT may be used to establish which API calls can be made and information in the context token may be used to determine
which specific resources can be accessed.

> The `Proxy-Authorization` header field ({{Section 11.7.2 of RFC9110}}) is a distinct field and is unaffected by the

Author: I like the idea that this change makes support for Proxy-Authorization much more viable. Curious to get thoughts on whether we should talk about it directly in this document and have security considerations on it?

Reviewer: I think I could get onboard with that idea but it seems like this paragraph says it's undefined but then goes on to kinda define it.

Author: Removed the second part of the paragraph and now only say that it's out of scope.

## Error Conditions

> Errors may occur during the processing of the WPT. If the signature verification fails for any reason,

Reviewer: ```suggestion
Errors may occur during the processing of the WPT. If the validation fails for any reason,
```
